README.TXT
Linux Software (DM35425)

SWP-700010163 rev D

Version v03.00.01.125638

Copyright (C) RTD Embedded Technologies, Inc.  All Rights Reserved.

This software package is dual-licensed.  Source code that is compiled for
kernel mode execution is licensed under the GNU General Public License
version 2.  For a copy of this license, refer to the file
LICENSE_GPLv2.TXT (which should be included with this software) or contact
the Free Software Foundation.  Source code that is compiled for user mode
execution is licensed under the RTD End-User Software License Agreement.
For a copy of this license, refer to LICENSE.TXT or contact RTD Embedded
Technologies, Inc.  Using this software indicates agreement with the
license terms listed above.


------------------
Supported Hardware
------------------

This software supports the RTD DM35425.


-------------------------
Supported Kernel Versions
-------------------------

This software has been tested with the following Linux distributions and kernel
versions:

    * openSUSE 12.2 (unmodified kernel 3.2)
    * Ubuntu 16.04 LTS (unmodified kernel 4.4)
    * Ubuntu 16.04 LTS (4.10 kernel built from kernel.org source)

 Due to API differences between kernel versions, RTD cannot guarantee compatibility
 with kernels and distributions not listed above.  If a user wishes to use an 
 unsupported kernel/distribution, it may be necessary to modify the driver module 
 code and/or Makefiles for the specific Linux environment.
 
--------------------------
Supported CPU Architecture
--------------------------
  
 This software has been validated on the following CPU architectures.
 
    * i586 (32-bit) dual-core
    * i586 (64-bit) single-core
    * i586 (64-bit) multi-core


-------------------
Supported Compilers
-------------------

The driver software and example programs were compiled using the GNU gcc 
compiler under openSUSE 12.2 and Ubuntu 16.04


-------------
Driver Design
-------------



        +------------------+        +----------------------------+
        | Example Programs |   OR   | User-Developed Application |
        | Provided by RTD  |        |                            |
        +------------------+        +----------------------------+
                 |                                |
                 |                                |
                 +--------------------------------+
                               |
                               | High-level C Functions
                               |
                          +---------+
                          | Library |
                          +---------+
                               |
                               | Low-level IOCTLs
    User Mode                  |
    ==========================================================================
    Kernel Mode                |
                           +--------+
                           | Driver |
                           +--------+
                               |
                               | Kernel Services
                               |
    ==========================================================================
                            Hardware
                            
                            
------
Driver
------

The directory driver/ contains source code related to the driver.


In order to use a driver, one must first compile it, load it into the kernel,
and create device files for the board(s).  To do this, issue the following
commands while sitting in the driver/ directory:

    * make
    * sudo make load



The driver module must be loaded before running any program which accesses a
DM35425 device.



-----------------
Library Interface
-----------------

The directory lib/ contains source code related to the user library.


The DM35425 library is created with a file name of librtd-dm35425.a and is
statically linked.


Please refer to the software manual for details on using the user level library
functions.  These functions are prototyped in the file include/dm35425_library.h;
this header file must be included in any code which wishes to call library
functions.


The library must be built before compiling the example programs or your
application.


To build the library, issue the command "make" within lib/.


------------
Header Files
------------

The directory include/ contains all header files needed by the driver, example
programs, library, and user applications.

-------------
Documentation
-------------

With doxygen and graphviz installed, execute `doxygen doxyconfig` in the root
of this repository.


----------------
Example Programs
----------------

The directory examples/ contains source code related to the example programs,
which demonstrate how to use features of the DM35425 boards, test the driver, or
test the library.  In addition to source files, examples/ holds other files as
well; the purpose of these files will be explained below.


To build the example programs, issue the command "make" within examples/.


The following files are provided in examples/:

    * Makefile
	    Make description file for building example programs.

	* dm35425_adc_multiboard_dma.c
		This example program demonstrates the use multiple ADCs that are driven
		at the same rate, where interrupts are handled by a single thread.

		The example program as-is requires 3 boards to operate, but can be
		minimally modified to support 1--n number of boards.

		Hit CTRL-C to exit.

    * dm35425_adc.c
            This example program demonstrates the use of the ADC and interrupt
            handling.  An interrupt is generated each time an ADC has taken a 
            sample.  When that interrupt happens, the programs gets the value
            of the last sample and displays it on the screen.
            
            Setup: Connect the signal of interest to AIN0 (pin 1 of CN3) and AGND
            (pin 21 of CN3).
            
            Usage: Display the command syntax by executing 
            
                           ./dm35425_adc --help.
            
            Hit CTRL-C to exit.

    * dm35425_adc_all_dma.c
            Demonstrate using the ADC and DMA, with all 32 channels active at
            the same time.  Data will be collected and written to files, 1 file
            per channel.
            
            Setup:  Connect the signal of interest to any analog input pin.
            
            Usage:  Display the command syntax by executing
            
                           ./dm35425_adc_all_dma --help

            Hit CTRL-C to exit.
            
    * dm35425_adc_continuous_dma.c
            This example program demonstrates the use of the ADC and DMA.  The 
            example will collect data from the ADC via DMA, and then write the
            data out to a file on disk.  The data can then be plotted using 
            gnuplot and the plot_adc_dma file.
            
            Setup: Connect the signal of interest to AIN0 (pin 1 of CN3) and AGND
            (pin 21 of CN3)
            
            Usage: Display the command syntax by executing 
            
                           ./dm35425_adc_continuous --help.
                
            Hit CTRL-C to exit.

    * dm35425_adc_fifo.c
            Demonstrates using the FIFO to collect data from the ADC, and 
            what interrupts occur while doing so.  The example will sample
            data enough to fill the FIFO, read the data from the FIFO and
            print it to the screen, and will trigger various interrupts.
            The example will complete and exit without user intervention.
            
            Setup: Connect the signal of interest to AIN0 (CN3 Pin 1)
            
            Usage: Display the command syntax by executing
                           ./dm35425_adc_fifo --help
                           
    * dm35425_adio.c
            This example program sets 16-bits of DIO to output, and
            16-bits to input.  We'll connect the output to the input and
            then write every possible 16-bit value to the output
            and verify the same value on the input pins.
            
            Setup: On CN3 and CN4, make these connections (using a standard jumper
            is usually the easiest way)
            
		Pin23 to Pin24
		Pin25 to Pin26
		Pin27 to Pin28
		Pin29 to Pin30
		Pin31 to Pin32
		Pin33 to Pin34
		Pin35 to Pin36
		Pin37 to Pin38            
            
            Usage: Display the command syntax by executing 
            
                           ./dm35425_adc_continuous --help.
                
            Hit CTRL-C to exit.

    * dm35425_adio_adv_int.c
            This example program uses the DIO match and event interrupts.
            It will match on the value 0xAA, and when that value passes
            through the input pins, it will throw an interrupt.

            The example will then wait for an event interrupt, which will occur
            if any input bit changes to zero.  This can be accomplished
            by touching a GND wire to any of the input ADIO pins.  Alternately,
            hitting Enter during the example will toggle the output bits,
            accomplishing the same thing.
            
            Setup: On CN3 and CN4, make these connections (using a standard jumper
            is usually the easiest way)
            
		Pin23 to Pin24
		Pin25 to Pin26
		Pin27 to Pin28
		Pin29 to Pin30
		Pin31 to Pin32
		Pin33 to Pin34
		Pin35 to Pin36
		Pin37 to Pin38  
            
            Usage: Display the command syntax by executing 
            
                           ./dm35425_adc_continuous --help.
                
            Hit CTRL-C to exit.

    * dm35425_adio_dma.c
            The example will make use of 3 DMA buffers for each
            of the three DMA channels (ADIO In, ADIO Out, and ADIO
            Direction).  Data will "play out" of the ADIO Out and
            Direction channels, and be stored in the ADIO In DMA
            buffer.  Doing this, we'll receive a pattern in the DMA
            In buffers that is the result of the output and changing
            bit direction values.
            
            Setup: On CN3 and CN4, make these connections (using a standard jumper
            is usually the easiest way)
            
		Pin23 to Pin24
		Pin25 to Pin26
		Pin27 to Pin28
		Pin29 to Pin30
		Pin31 to Pin32
		Pin33 to Pin34
		Pin35 to Pin36
		Pin37 to Pin38  
            
            Usage: Display the command syntax by executing 
            
                           ./dm35425_adc_continuous --help.
                
            Hit CTRL-C to exit.

    * dm35425_adio_parallel_bus.c
            The ADIO may be used as a parallel bus to transfer data from 1 board
	    to another.  In this mode, 3 ADIO signals are used for control, and
	    the remaining 29 bits are used for passing data. This board uses DMA
            and the parallel bus mode to transfer data from 1 board to another.
            
            Setup: All ADIO pins on CN3 (Pins 23-38) must be connected from 1 board
	    to CN3 (Pins 23-38) of the 2nd board.

	    The three control lines on CN4 must also be connected between the
	    boards:

	    CN4 Pin 24
	    CN4 Pin 26
	    CN4 Pin 28
            
            Usage: Display the command syntax by executing 
            
                           ./dm35425_adc_continuous --help.
                
            Hit CTRL-C to exit.

    * dm35425_dac.c
            This example program demonstrates the use of the DAC.  A voltage is
            put out on the pin corresponding to the input from the user.  The voltage
            can be easily changed to cover the full range.
            
            Setup: Connect an oscilloscope to the AOUT0 pin (CN3 Pin 17).
            
            Usage: Display the command syntax by executing
            
                       ./dm35425_dac --help.
            
            Follow the prompts on screen for changing the voltage.
            
            Hit CTRL-C to exit.

    * dm35425_dac_dma.c
            This example program demonstrates the use of the DAC and DMA.  A 
            wave pattern data is generated in the program, written to a DMA buffer,
            and then sent to the DAC in a repeating loop, thus provding a
            continuous cycling signal.
            
            Setup: Connect an oscilloscope to the AOUT0 pin (CN3 Pin 17).
            
            Usage: Display the command syntax by executing
            
                      ./dm35425_dac_dma --help.
            
            Hit CTRL-C to exit.          
            
    * dm35425_ext_clocking.c
            This example program uses function blocks to create signals which
	    are looped back into external clock inputs.  Each signal generated
            produces the equivalent of a square wave.
            
            The output shows that each function block runs at half the speed of
            the previous one due to the cascading clocks.
            
            Setup: Make connections as follows:

            CN3: Pin 17 to Pin 39
            CN3: Pin 37 to Pin 40
            
            Usage: Display the command syntax by executing 
            
                      ./dm35425_ext_clocking --help.
                      
            Hit CTRL-C to exit.                      
            
    * dm35425_list_fb.c
            This example program demonstrates using the board-level 
            registers to access the function blocks on the board.  The 
            program will query every function block location to see if a 
            valid function block type exists there, and if it does, it will
            display that type on the screen.  In this way, it will give an
            inventory of the function blocks on the board.
            
            Setup: No setup required.
            
            Usage: ./dm35425_list_fb --minor 0



-----------------
Known Limitations

CHANGES.TXT
Change log for SWP-700010163 rev D

Linux Software (DM35425)

Copyright (C) RTD Embedded Technologies, Inc.  All Rights Reserved.


v01.00.00 (rev A)   8/7/2014  RGG
-----------------------------------
- Initial version


v02.00.00 (rev B)   3/20/2016  RGG
-----------------------------------
- Explicitly calling pci_set_master in driver to allow DMA, as some
  UEFI BIOS do not set this in PCI space.
- Removed library function _Find_Used_Buffer because it had some inherent flaws
  and in very remote cases could fail.  The user is now responsible to keep
  track of what buffer is next.  All examples using DMA have been updated to
  demonstrate how to do this.
- Changed method in library for setting sample/conversion rate.  It now gets the 
  system clock from another library call (which constrains it) instead of reading
  it directly.  This solves a problem where consecutive calls to set the same
  sample rate was resulting in different rates for different function blocks.
- Added adc_fifo examples to show using the FIFO for data transfer without DMA.
- Added adc_all_dma, showing use of all of the ADC channels simultaneously.
- Removed library function for setting ISR priority.  Using it required being root,
  and it's not actually needed anyway.
  

v03.00.00 (rev C)    05/16/2017     RGG
---------------------------------------
- Added revision info from GBC registers to list_fb example output.
- DMA Buffer setup library call changed.  It now takes a single unsigned
  8-bit value for the register instead of booleans for every bit.
- Dma Initialize and other functions moved out of DMA library and into 
  os-specific library.
- Added reference count to driver.  This prevents the device file from being
  opened while already held open by another program.
- All instances of CLK_SRC_GBL have been changed to CLK_BUS for clarity.
- Removed setting rate in _dac example, as it is not needed.
- Example program adc_continuous_dma did not work correctly.


v03.00.01 (rev D)  03/05/2020  LRF
-----------------------------------
- Added Support for the Linux 5.0 Kernel Module make.
- Added support for 5.0+ kernel module dma functions.  
- Added support for 5.6+ Kernel, Removed the ioremap_nocache(...) and replaced with
its nondeprecated version ioremap(...) that accomplishes the same thing.

v03.00.01 (rev D1) 05/15/2023 SKM
-----------------------------------
- Git-ified the repository.

v03.01.00 (rev E) 05/18/2023 SKM
-----------------------------------
- Added ADC Multi-board Readout API.

examples/dm35425_adc.c
/**
    @file

    @brief
        Example program which demonstrates the use of the ADC, setting
        and responding to interrupts.

    @verbatim

        This example program uses an ADC to collect data.  An interrupt
        is generated every time data is collected by the ADC.
        After acknowledging the interrupt, the program queries the value
        last taken by the ADC, and the sample counter, and prints them
        to the screen.

	    Connect the signal of interest to AIN0 (CN3 Pin 1) and AGND
        (CN3 Pin 21), or pins corresponding to selected channel.

        For convenience in testing the ADC, especially differential voltages,
        the DAC is setup to output these specific voltages:

        AOUT0: -6V
        AOUT1: -3V
        AOUT2:  4V
        AOUT3:  8V

        The program will continue to run until CTRL-C is pressed.

    @endverbatim

    @verbatim
    --------------------------------------------------------------------------
    This file and its contents are copyright (C) RTD Embedded Technologies,
    Inc.  All Rights Reserved.

    This software is licensed as described in the RTD End-User Software License
    Agreement.  For a copy of this agreement, refer to the file LICENSE.TXT
    (which should be included with this software) or contact RTD Embedded
    Technologies, Inc.
    --------------------------------------------------------------------------
    @endverbatim

    $Id: dm35425_adc.c 108025 2017-04-14 15:09:34Z rgroner $
*/

#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <errno.h>
#include <error.h>
#include <unistd.h>
#include <limits.h>
#include <signal.h>
#include <getopt.h>
#include <signal.h>
#include <string.h>

#include "dm35425_gbc_library.h"
#include "dm35425_adc_library.h"
#include "dm35425_dac_library.h"
#include "dm35425_ioctl.h"
#include "dm35425_examples.h"
#include "dm35425_util_library.h"
#include "dm35425_board_access.h"
#include "dm35425_types.h"
#include "dm35425.h"
#include "dm35425_os.h"

/**
 * Rate to run at, if the user does not provide one.
 */
#define DEFAULT_RATE	1000

/**
 * Define a channel to use, if the user does not
 * provide one.
 */
#define DEFAULT_CHANNEL		0

/**
 * Define a default range to use, if the user does not
 * provide one.
 */
#define DEFAULT_RANGE		DM35425_ADC_RNG_BIPOLAR_5V

/**
 * Define a default mode to use, if the user does not provide one
 */
#define DEFAULT_MODE		DM35425_ADC_INPUT_SINGLE_ENDED

/**
 * Name of the program as invoked on the command line
 */
static char *program_name;

/**
 * Count of interrupts that have happened.
 */
volatile int interrupt_count = 0;

/**
 * Boolean indicating whether or not to exit the program.
 */
volatile int exit_program = 0;

/**
*******************************************************************************
@brief
    Print information on stderr about how the program is to be used.  After
    doing so, the program is exited.
 *******************************************************************************
*/

static void usage(void)
{
	fprintf(stderr, "\n");
	fprintf(stderr, "NAME\n\n\t%s\n\n", program_name);
	fprintf(stderr, "USAGE\n\n\t%s [OPTIONS]\n\n", program_name);

	fprintf(stderr, "OPTIONS\n\n");

	fprintf(stderr, "\t--help\n");
	fprintf(stderr, "\t\tShow this help screen and exit.\n");

	fprintf(stderr, "\t--minor NUM\n");
	fprintf(stderr,
		"\t\tSpecify the minor number (>= 0) of the board to open.  When not specified,\n");
	fprintf(stderr, "\t\tthe device file with minor 0 is opened.\n");

	fprintf(stderr, "\t--channel CHAN\n");
	fprintf(stderr,
		"\t\tUse the specified ADC input channel.  Defaults to channel 0.\n");

	fprintf(stderr, "\t--range RNG\n");
	fprintf(stderr, "\t\tUse the specified range and mode of the ADC.\n");
	fprintf(stderr, "\t\t\t10B = 10V, Bipolar\n");
	fprintf(stderr, "\t\t\t10U = 10V, Unipolar\n");
	fprintf(stderr, "\t\t\t5B = 5V, Bipolar (Default)\n");
	fprintf(stderr, "\t\t\t5U = 5V, Unipolar\n");
	fprintf(stderr, "\t\t\t2.5B = 2.5V, Bipolar\n");
	fprintf(stderr, "\t\t\t2.5U = 2.5V, Unipolar\n");
	fprintf(stderr, "\t\t\t1.25B = 1.25V, Bipolar\n");
	fprintf(stderr, "\t\t\t1.25U = 1.25V, Unipolar\n");
	fprintf(stderr, "\t\t\t.625B = 0.625V, Bipolar\n");

	fprintf(stderr, "\t--mode MODE\n");
	fprintf(stderr, "\t\tChange the mode of the ADC.\n");
	fprintf(stderr, "\t\t\tse = single-ended (Default)\n");
	fprintf(stderr, "\t\t\tdiff = differential\n");

	fprintf(stderr, "\n");

	exit(EXIT_FAILURE);
}

/**
*******************************************************************************
@brief
    The interrupt subroutine that will execute when an interrupt occurs.  It will
    simply increment a count, which the main program
 *******************************************************************************
*/

void ISR(struct dm35425_ioctl_interrupt_info_request int_info)
{

	if (int_info.error_occurred) {

		printf("ISR: Error received.\n");
		return;

	}

	if (int_info.valid_interrupt) {

		interrupt_count++;

	}

}

/**
*******************************************************************************
@brief
    Signal handler for SIGINT Control-C keyboard interrupt.

@param
    signal_number

    Signal number passed in from the kernel.

@warning
    One must be extremely careful about what functions are called from a signal
    handler.
 *******************************************************************************
*/
static void sigint_handler(int signal_number)
{
	exit_program = 0xff;
}

/**
*******************************************************************************
@brief
    Prepare the DACs for use

@param
    board

    Pointer to the board descriptor

 *******************************************************************************
*/
static void DM35425_Setup_Dacs(struct DM35425_Board_Descriptor *board)
{

	struct DM35425_Function_Block my_dac;
	int result;
	int16_t conv_value = 0;

	/****************************
	 * Set DAC 0 Chan 0 to -6.0 volts
	 ****************************/
	result = DM35425_Dac_Open(board, DAC_0, &my_dac);

	check_result(result, "Could not open DAC");

	result = DM35425_Dac_Channel_Setup(board,
					   &my_dac,
					   CHANNEL_0,
					   DM35425_DAC_RNG_BIPOLAR_10V);

	check_result(result, "Error setting output range.");

	result = DM35425_Dac_Reset(board, &my_dac);

	check_result(result, "Error stopping DAC");

	result = DM35425_Dac_Volts_To_Conv(DM35425_DAC_RNG_BIPOLAR_10V,
					   -6.0f, &conv_value);

	check_result(result, "Error converting voltage to conversion");

	result = DM35425_Dac_Set_Last_Conversion(board,
						 &my_dac,
						 CHANNEL_0, 0, conv_value);
	check_result(result, "Error setting last conversion");

	/****************************
	 * Set DAC 0 Chan 1 to -3.0 volts
	 ****************************/
	result = DM35425_Dac_Channel_Setup(board,
					   &my_dac,
					   CHANNEL_1,
					   DM35425_DAC_RNG_BIPOLAR_10V);

	check_result(result, "Error setting output range.");

	result = DM35425_Dac_Reset(board, &my_dac);

	check_result(result, "Error stopping DAC");

	result = DM35425_Dac_Volts_To_Conv(DM35425_DAC_RNG_BIPOLAR_10V,
					   -3.0f, &conv_value);

	check_result(result, "Error converting voltage to conversion");

	result = DM35425_Dac_Set_Last_Conversion(board,
						 &my_dac,
						 CHANNEL_1, 0, conv_value);
	check_result(result, "Error setting last conversion");

	/****************************
	 * Set DAC 0 Chan 2 to 4.0 volts
	 ****************************/
	result = DM35425_Dac_Channel_Setup(board,
					   &my_dac,
					   CHANNEL_2,
					   DM35425_DAC_RNG_BIPOLAR_10V);

	check_result(result, "Error setting output range.");

	result = DM35425_Dac_Reset(board, &my_dac);

	check_result(result, "Error stopping DAC");

	result = DM35425_Dac_Volts_To_Conv(DM35425_DAC_RNG_BIPOLAR_10V,
					   4.0f, &conv_value);

	check_result(result, "Error converting voltage to conversion");

	result = DM35425_Dac_Set_Last_Conversion(board,
						 &my_dac,
						 CHANNEL_2, 0, conv_value);
	check_result(result, "Error setting last conversion");

	/****************************
	 * Set DAC 0 Chan 3 to 8.0 volts
	 ****************************/
	result = DM35425_Dac_Open(board, DAC_0, &my_dac);

	check_result(result, "Could not open DAC");

	result = DM35425_Dac_Channel_Setup(board,
					   &my_dac,
					   CHANNEL_3,
					   DM35425_DAC_RNG_BIPOLAR_10V);

	check_result(result, "Error setting output range.");

	result = DM35425_Dac_Reset(board, &my_dac);

	check_result(result, "Error stopping DAC");

	result = DM35425_Dac_Volts_To_Conv(DM35425_DAC_RNG_BIPOLAR_10V,
					   8.0f, &conv_value);

	check_result(result, "Error converting voltage to conversion");

	result = DM35425_Dac_Set_Last_Conversion(board,
						 &my_dac,
						 CHANNEL_3, 0, conv_value);
	check_result(result, "Error setting last conversion");

}

/**
*******************************************************************************
@brief
    The main program.

@param
    argument_count

    Number of args passed on the command line, including the executable name

@param
    arguments

    Pointer to array of character strings, which are the args themselves.

@retval
    0

    Success

@retval
    Non-Zero

    Failure.

 *******************************************************************************
*/
int main(int argument_count, char **arguments)
{
	struct DM35425_Board_Descriptor *board;
	struct DM35425_Function_Block my_adc;
	unsigned long int minor = 0;
	int result;
	int my_value;
	uint32_t sample_counts;
	int last_int_count = 0;
	uint16_t int_status;

	unsigned int actual_rate;
	unsigned int channel = DEFAULT_CHANNEL;
	unsigned int range = DEFAULT_RANGE;
	enum DM35425_Input_Mode mode = DEFAULT_MODE;

	int help_option_given = 0;
	int status;
	struct sigaction signal_action;

	uint16_t interrupt_ena = 0;
	float volts = 0;

	char *invalid_char_p;
	struct option options[] = {
		{"help", 0, 0, HELP_OPTION},
		{"minor", 1, 0, MINOR_OPTION},
		{"channel", 1, 0, CHANNELS_OPTION},
		{"range", 1, 0, RANGE_OPTION},
		{"mode", 1, 0, MODE_OPTION},
		{0, 0, 0, 0}
	};

	program_name = arguments[0];

	// Show usage, parse arguments
	while (1) {
		/*
		 * Parse the next command line option and any arguments it may require
		 */
		status = getopt_long(argument_count,
				     arguments, "", options, NULL);

		/*
		 * If getopt_long() returned -1, then all options have been processed
		 */
		if (status == -1) {
			break;
		}

		/*
		 * Figure out what getopt_long() found
		 */
		switch (status) {

			/*#################################################################
			   User entered '--help'
			   ################################################################# */
		case HELP_OPTION:
			help_option_given = 0xFF;
			break;

			/*#################################################################
			   User entered '--minor'
			   ################################################################# */
		case MINOR_OPTION:
			/*
			 * Convert option argument string to unsigned long integer
			 */
			errno = 0;
			minor = strtoul(optarg, &invalid_char_p, 10);

			/*
			 * Catch unsigned long int overflow
			 */
			if ((minor == ULONG_MAX)
			    && (errno == ERANGE)) {
				error(0, 0,
				      "ERROR: Device minor number caused numeric overflow");
				usage();
			}

			/*
			 * Catch argument strings with valid decimal prefixes, for
			 * example "1q", and argument strings which cannot be converted,
			 * for example "abc1"
			 */
			if ((*invalid_char_p != '\0')
			    || (invalid_char_p == optarg)) {
				error(0, 0,
				      "ERROR: Non-decimal device minor number");
				usage();
			}

			break;

			/*#################################################################
			   User entered '--channel'
			   ################################################################# */
		case CHANNELS_OPTION:
			/*
			 * Convert channel argument string to unsigned long integer
			 */
			errno = 0;
			channel = strtoul(optarg, &invalid_char_p, 10);

			/*
			 * Catch unsigned long int overflow
			 */
			if ((channel == ULONG_MAX)
			    && (errno == ERANGE)) {
				error(0, 0,
				      "ERROR: Channel number caused numeric overflow");
				usage();
			}

			/*
			 * Catch argument strings with valid decimal prefixes, for
			 * example "1q", and argument strings which cannot be converted,
			 * for example "abc1"
			 */
			if ((*invalid_char_p != '\0')
			    || (invalid_char_p == optarg)) {
				error(0, 0,
				      "ERROR: Non-decimal device minor number");
				usage();
			}

			if (channel >= DM35425_NUM_ADC_DMA_CHANNELS) {
				error(0, 0,
				      "ERROR: Channel number must be from 0 to %u.",
				      DM35425_NUM_ADC_DMA_CHANNELS - 1);
				usage();
			}
			break;

			/*#################################################################
			   User entered --range option
			   ################################################################# */
		case RANGE_OPTION:
			if (strcmp(optarg, "10B") == 0) {
				range = DM35425_ADC_RNG_BIPOLAR_10V;
			} else if (strcmp(optarg, "10U") == 0) {
				range = DM35425_ADC_RNG_UNIPOLAR_10V;
			} else if (strcmp(optarg, "5B") == 0) {
				range = DM35425_ADC_RNG_BIPOLAR_5V;
			} else if (strcmp(optarg, "5U") == 0) {
				range = DM35425_ADC_RNG_UNIPOLAR_5V;
			} else if (strcmp(optarg, "2.5B") == 0) {
				range = DM35425_ADC_RNG_BIPOLAR_2_5V;
			} else if (strcmp(optarg, "2.5U") == 0) {
				range = DM35425_ADC_RNG_UNIPOLAR_2_5V;
			} else if (strcmp(optarg, "1.25U") == 0) {
				range = DM35425_ADC_RNG_UNIPOLAR_1_25V;
			} else if (strcmp(optarg, "1.25B") == 0) {
				range = DM35425_ADC_RNG_BIPOLAR_1_25V;
			} else if (strcmp(optarg, ".625B") == 0) {
				range = DM35425_ADC_RNG_BIPOLAR_625mV;
			} else {
				error(0, 0,
				      "ERROR: Range and mode entered did not match available options.");
				usage();
			}

			break;

			/*#################################################################
			   User entered --mode option
			   ################################################################# */
		case MODE_OPTION:
			if (strcmp(optarg, "se") == 0) {
				mode = DM35425_ADC_INPUT_SINGLE_ENDED;
			} else if (strcmp(optarg, "diff") == 0) {
				mode = DM35425_ADC_INPUT_DIFFERENTIAL;
			} else {
				error(0, 0,
				      "ERROR: Mode must be either se or diff.");
				usage();
			}
			break;

			/*#################################################################
			   User entered unsupported option
			   ################################################################# */
		case '?':
			usage();
			break;

			/*#################################################################
			   getopt_long() returned unexpected value
			   ################################################################# */
		default:
			error(EXIT_FAILURE,
			      0,
			      "ERROR: getopt_long() returned unexpected value %#x",
			      status);
			break;
		}
	}

	/*
	 * Recognize '--help' option before any others
	 */

	if (help_option_given) {
		usage();
	}

	signal_action.sa_handler = sigint_handler;
	sigfillset(&(signal_action.sa_mask));
	signal_action.sa_flags = 0;

	if (sigaction(SIGINT, &signal_action, NULL) < 0) {
		error(EXIT_FAILURE, errno, "ERROR: sigaction() FAILED");
	}

	printf("Opening board.....");
	result = DM35425_Board_Open(minor, &board);

	check_result(result, "Could not open board");
	printf("success.\nResetting board.....");
	result = DM35425_Gbc_Board_Reset(board);

	check_result(result, "Could not reset board");
	printf("success.\n");

	DM35425_Setup_Dacs(board);

	printf("Opening ADC......\n");

	result = DM35425_Adc_Open(board, ADC_0, &my_adc);

	check_result(result, "Could not open ADC");

	printf("Found ADC, with %d DMA channels (%d buffers each)\n",
	       my_adc.num_dma_channels, my_adc.num_dma_buffers);

	printf("Using Channel %u\n", channel);

	result = DM35425_Adc_Set_Clock_Src(board,
					   &my_adc, DM35425_CLK_SRC_IMMEDIATE);

	check_result(result, "Error setting ADC clock");

	result = DM35425_Adc_Set_Pre_Trigger_Samples(board, &my_adc, 0);

	check_result(result, "Error setting pre-capture samples.");

	result = DM35425_Adc_Set_Post_Stop_Samples(board, &my_adc, 0);

	check_result(result, "Error setting post-capture samples.");

	fprintf(stdout, "Installing user ISR ...\n");
	result = DM35425_General_InstallISR(board, ISR);
	check_result(result, "DM35425_General_InstallISR()");

	result = DM35425_Adc_Interrupt_Set_Config(board,
						  &my_adc,
						  DM35425_ADC_INT_SAMPLE_TAKEN_MASK,
						  INTERRUPT_ENABLE);

	check_result(result, "Error setting interrupt.");

	result = DM35425_Adc_Channel_Setup(board,
					   &my_adc,
					   channel,
					   DM35425_ADC_NO_DELAY, range, mode);

	check_result(result, "Error setting up channel.");

	printf("Initializing ADC......\n");
	result = DM35425_Adc_Set_Start_Trigger(board,
					       &my_adc,
					       DM35425_CLK_SRC_IMMEDIATE);
	check_result(result, "Error setting start trigger.");

	result = DM35425_Adc_Set_Stop_Trigger(board,
					      &my_adc, DM35425_CLK_SRC_NEVER);
	check_result(result, "Error setting stop trigger.");

	result = DM35425_Adc_Set_Sample_Rate(board,
					     &my_adc,
					     DEFAULT_RATE, &actual_rate);

	check_result(result, "Failed to set sample rate for ADC.");
	fprintf(stdout,
		"ADC0: Rate requested: %d  Actual Rate Achieved: %d\n",
		DEFAULT_RATE, actual_rate);

	result = DM35425_Adc_Initialize(board, &my_adc);

	check_result(result, "Failed or timed out initializing ADC.");

	result = DM35425_Adc_Start(board, &my_adc);

	check_result(result, "Error starting ADC");

	result = DM35425_Adc_Interrupt_Get_Config(board,
						  &my_adc, &interrupt_ena);

	check_result(result, "Error getting interrupt value");

	printf("\n\nPress Ctrl-C to exit.\n\n");

	printf("Sample Count\tVoltage   \tADC Counts\n");
	printf("============\t==========\t=============\n");

	while (!exit_program) {
		if (last_int_count < interrupt_count) {

			result = DM35425_Adc_Interrupt_Get_Status(board,
								  &my_adc,
								  &int_status);
			check_result(result, "Error getting interrupt status");

			result = DM35425_Adc_Get_Sample_Count(board,
							      &my_adc,
							      &sample_counts);

			check_result(result, "Error getting sample count.");

			printf("%12d", sample_counts);

			result =
			    DM35425_Adc_Channel_Get_Last_Sample(board,
								&my_adc,
								channel,
								&my_value);

			check_result(result, "Error getting ADC value.");

			result =
			    DM35425_Adc_Sample_To_Volts(range,
							my_value, &volts);

			check_result(result,
				     "Error converting ADC sample to volts.");

			printf("\t%+2.5f\t%d    \r", volts, my_value);

			result = DM35425_Adc_Interrupt_Clear_Status(board,
								    &my_adc,
								    int_status);
			check_result(result, "Error clearing interrupt status");

			last_int_count++;
		}

		DM35425_Micro_Sleep(100);
	}

	printf("\n\nStopping Adc............");

	result = DM35425_Adc_Reset(board, &my_adc);

	check_result(result, "Error starting ADC");

	printf("success!\nDisabling interrupt.....");

	result = DM35425_Adc_Interrupt_Set_Config(board,
						  &my_adc,
						  DM35425_ADC_INT_SAMPLE_TAKEN_MASK,
						  INTERRUPT_DISABLE);

	check_result(result, "Error removing interrupt.");

	printf("success!\nRemoving ISR......");
	result = DM35425_General_RemoveISR(board);

	check_result(result, "Error removing ISR.");

	printf("success.\n");

	printf("Closing Board\n");
	result = DM35425_Board_Close(board);

	check_result(result, "Error closing board.");
	printf("Example program successfully completed.\n");
	return 0;

}
